### PR TITLE
Allow scrolling right-to-left or bottom-to-top scrollable areas by dragging the scrollbar.

### PR DESCRIPTION
**Pull request by GetToSet**
#### 8074107aa32e4b5d15c17939e7ae877a03a40f8d
<pre>
Allow scrolling right-to-left or bottom-to-top scrollable areas by dragging the scrollbar.
<a href="https://bugs.webkit.org/show_bug.cgi?id=266555">https://bugs.webkit.org/show_bug.cgi?id=266555</a>

Reviewed by NOBODY (OOPS!).

ScrollableArea::scrollToOffsetWithoutAnimation(ScrollbarOrientation, float offset)
should convert the offsets into scroll positions before calling
scrollToPositionWithoutAnimation, hence taking the scrollOrigin into consideration.

This fixes issues that right-to-left or bottom-to-top scrollable areas are not
scrollable by dragging the scrollbar.

* LayoutTests/fast/scrolling/rtl-bt-drag-scroller-expected.txt: Added.
* LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::scrollToPositionWithoutAnimation):
(WebCore::ScrollableArea::scrollToOffsetWithoutAnimation):
* Source/WebCore/platform/ScrollableArea.h:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::layout):
(WebCore::RenderListBox::scrollToRevealElementAtListIndex):
(WebCore::RenderListBox::setLogicalScrollTop):
</pre>
https://github.com/WebKit/WebKit/commit/f80b57870337763f95c4f22bc96cce3dd813dc91

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31357 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10026 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33050 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33851 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28423 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7276 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28063 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8443 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27996 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7265 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7428 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35192 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28507 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33560 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7493 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31399 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9156 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27703 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8186 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8004 "Built successfully") | | |

### CODE REVIEW

File: LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html [hunk 1]
```diff
@@ -0,0 +1,31 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="../../resources/js-test-pre.js"></script>
```
Comment (aprotyas) [inline LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html]: You can just use `js-test.js` here and remove the `-post.js` file from the bottom.

File: Source/WebCore/platform/ScrollableArea.cpp [hunk 2]
```diff
@@ -172,6 +172,15 @@ void ScrollableArea::scrollToPositionWithoutAnimation(const FloatPoint& position
     scrollAnimator().scrollToPositionWithoutAnimation(position, clamping);
 }
 
+void ScrollableArea::scrollToPositionWithoutAnimation(ScrollbarOrientation orientation, float offset)
```
Comment (aprotyas) [inline Source/WebCore/platform/ScrollableArea.cpp]: Since this function isn't about scrolling to an offset, consider changing the argument name:
```suggestion
void ScrollableArea::scrollToPositionWithoutAnimation(ScrollbarOrientation orientation, float position)
```

**Review by aprotyas — COMMENTED** [review]
Editorial comment about the change log. Please remove this line:

```diff
--- a	2023-12-17 10:31:21
+++ b	2023-12-17 10:31:26
@@ -10,7 +10,6 @@
 This fixes issues that right-to-left or bottom-to-top scrollable areas are not
 scrollable by dragging the scrollbar.

-Combined changes:
 * LayoutTests/fast/scrolling/rtl-bt-drag-scroller-expected.txt: Added.
 * LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html: Added.
 * Source/WebCore/platform/ScrollableArea.cpp:
```

File: LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+<body style="margin: 0px;">
+<div id="scrollMe" style="writing-mode: vertical-rl; direction: rtl; width: 100px; height: 100px; position: absolute; left: 0px; top: 0px; overflow: scroll;">
+<div style="width: 300px; height: 300px; background-color: yellow;"></div>
+</div>
+<script>
+description("This test makes sure that a div with right-to-left or bottom-to-top scroll directions can be scrolled by dragging the scrollbar.");
+
+if (window.eventSender) {
+    eventSender.mouseMoveTo(96, 96);
+    eventSender.mouseDown();
```
Comment (Ahmad-S792) [inline LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html:15]: I think this will fail on `iOS`. @pxlcoder - Does the test need to be using 'UIHelper' to make it work on 'iOS' with 'scrollTo' usage?

Comment (GetToSet) [reply LayoutTests/fast/scrolling/rtl-bt-drag-scroller.html:15]: I've added a expectation for failure on iOS. Meanwhile, I found this issue reproducible on WebKitGTK as well so it's not Mac-only.

File: Source/WebCore/rendering/RenderListBox.cpp [hunk 4]
```diff
@@ -731,7 +731,7 @@ bool RenderListBox::scrollToRevealElementAtListIndex(int index)
     if (style().isFlippedBlocksWritingMode())
         newOffset *= -1;
 
-    scrollToOffsetWithoutAnimation(scrollbarOrientationForWritingMode(), newOffset);
+    scrollToPositionWithoutAnimation(scrollbarOrientationForWritingMode(), newOffset);
```
Comment (smfr) [inline Source/WebCore/rendering/RenderListBox.cpp]: The function says "position" but the argument says "offset".

File: Source/WebCore/platform/ScrollableArea.cpp [hunk 5]
```diff
@@ -207,9 +216,9 @@ void ScrollableArea::scrollToOffsetWithoutAnimation(ScrollbarOrientation orienta
 {
     auto currentPosition = scrollAnimator().currentPosition();
     if (orientation == ScrollbarOrientation::Horizontal)
-        scrollAnimator().scrollToPositionWithoutAnimation(FloatPoint(offset, currentPosition.y()));
+        scrollAnimator().scrollToPositionWithoutAnimation(FloatPoint(offset - scrollOrigin().x(), currentPosition.y()));
```
Comment (smfr) [inline Source/WebCore/platform/ScrollableArea.cpp]: Please rewrite to use the helper `scrollPositionFromOffset()`.

File: Source/WebCore/platform/ScrollableArea.cpp [hunk 6]
```diff
@@ -207,9 +216,9 @@ void ScrollableArea::scrollToOffsetWithoutAnimation(ScrollbarOrientation orienta
...
     else
-        scrollAnimator().scrollToPositionWithoutAnimation(FloatPoint(currentPosition.x(), offset));
+        scrollAnimator().scrollToPositionWithoutAnimation(FloatPoint(currentPosition.x(), offset - scrollOrigin().y()));
```
Comment (smfr) [inline Source/WebCore/platform/ScrollableArea.cpp]: Ditto.

File: Source/WebCore/platform/ScrollableArea.cpp (hunk 5 above)
Comment (smfr) [reply Source/WebCore/platform/ScrollableArea.cpp]: I'm not sure the addition of `scrollToPositionWithoutAnimation(, float)` is justified, especially since it just does the `if (orientation == ScrollbarOrientation::Horizontal)` check again.

**Review by smfr — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/ScrollableArea.cpp (hunk 5 above)
Comment (GetToSet) [reply Source/WebCore/platform/ScrollableArea.cpp]: > I'm not sure the addition of `scrollToPositionWithoutAnimation(, float)` is justified, especially since it just does the `if (orientation == ScrollbarOrientation::Horizontal)` check again.

Instead of scrollbar usages, existing usages at `RenderListBox` are passing positions instead of scrolling offsets to the function and it would be more intuitive since flipped writing modes are already handled.

File: Source/WebCore/platform/Scrollbar.cpp [hunk 7]
```diff
@@ -286,7 +286,7 @@ void Scrollbar::moveThumb(int pos, bool draggingDocument)
             destinationPosition = std::min(destinationPosition + delta, maximum());
```
Comment (GetToSet) [inline Source/WebCore/platform/Scrollbar.cpp:286]: The `draggingDocument` path are calling with positions, so the minimum and maximum values should consider `scrollOrigin`s when clamping. This should be fixed with a separate patch.
